### ISSUE TITLE
Issue with cupsCreateDestJob and raw print queues

### ISSUE THREAD

**Issue by ProgrammingPescador** [apple/cups#5235]
Hello

I am trying to write a small application that can send ZPL commands to a printer connected to a Raspberry Pi.  I am using the examples in https://www.cups.org/doc/cupspm.html#introduction and am having some issues!

I have used `cups_dest_t *raw = cupsGetDest("Raw", NULL, count, dests);` to get my destination which returns what I believe to be the correct data, and then, `cups_dinfo_t *info = cupsCopyDestInfo(CUPS_HTTP_DEFAULT, raw);` to get the info pointer the pointer is returned as `0x0`.  Is this a problem?

I then use `ipp_status_t status = cupsCreateDestJob(CUPS_HTTP_DEFAULT, raw, info, &job_id, "Test Label", num_options, options);` to create a job and this returns `IPP_STATUS_ERROR_INTERNAL`

What is going wrong?

the full listing is:
`#include <stdio.h>
#include <cups/cups.h>

typedef struct
{
	int num_dests;
	cups_dest_t *dests;
} my_user_data_t;

int
my_dest_cb(my_user_data_t *user_data, unsigned flags, cups_dest_t *dest)
{
	if (flags & CUPS_DEST_FLAGS_REMOVED)
	{
		/*
		* Remove destination from array...
		*/

		user_data->num_dests =
			cupsRemoveDest(dest->name, dest->instance,
				user_data->num_dests,
				&(user_data->dests));
	}
	else
	{
		/*
		* Add destination to array...
		*/

		user_data->num_dests =
			cupsCopyDest(dest, user_data->num_dests,
				&(user_data->dests));
	}

	return (1);
}

int my_get_dests(cups_ptype_t type, cups_ptype_t mask, cups_dest_t **dests)
{
	my_user_data_t user_data = { 0, NULL };

	if (!cupsEnumDests(CUPS_DEST_FLAGS_NONE, 1000, NULL, type,
		mask, (cups_dest_cb_t)my_dest_cb,
		&user_data))
	{
		/*
		* An error occurred, free all of the destinations and
		* return...
		*/

		cupsFreeDests(user_data.num_dests, user_data.dests);

		*dests = NULL;

		return (0);
	}

	/*
	* Return the destination array...
	*/

	*dests = user_data.dests;

	return (user_data.num_dests);
}

int main(void)
{
	cups_dest_t *dests;

	int count = my_get_dests(NULL, NULL, &dests);

	if (count > 0)
	{
		cups_dest_t *raw = cupsGetDest("Raw", NULL, count, dests);
		if (raw)
		{
			printf("%s\n", raw->name);

			int job_id = 0;
			int num_options = 0;
			cups_option_t *options = NULL;
			cups_dinfo_t *info = cupsCopyDestInfo(CUPS_HTTP_DEFAULT, raw);

			num_options = cupsAddOption(CUPS_COPIES, "1", num_options, &options);

			ipp_status_t status = cupsCreateDestJob(CUPS_HTTP_DEFAULT, raw, info, &job_id, "Test Label", num_options, options);
			if (status == IPP_STATUS_OK)
			{
				char buffer[] = "^XA^FO300,50^ABN,22,14^FDCloudFoot^FS^FO300,100^GFA,1300,1300,13,00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000038000000000000000000000000CC0000000000000000000000010600000000000000000000000202000000000000000000000004010000000000000000000000080100000000000000000000001800800000000000000000000010009FFC0000000000000000001000F003000000000000000001A000C001800000000000000003600000004000000000000000022000000020000000000000003E000000003FC0000000000001C00000000010600000000000060000000000001000000000000C000000000000080000000000100000000000000800000000001000000000000004000000000010000000000000040000000000600000000000000C0000000000C00000000000000E0000000001000000000000000180000000030000000000000000C00000000200000000000000006000000002000000000000000030000000020000000000000000100000000200000000000000001000000002000000000000000018000000030000000000000000080000000180000000000000000800000000C0000000000000000800000000700000000000000008000000000F800000000000000800000000008000000000000010000000000080000000000000100000000000800000000000306000000000008000000000002FC0000000000040000000200040000000000000600000003000C00000000000003000060078018000000000000018001E004E06000000000000000700790047F80000000000000001FFE180830000000000000000003E20C30300000000000000000000207E01000000000000000000002000010000000000000000000040000100000000000000000001C00021000000000000000000070008210000000000000000003C000421000000000000000003C00006618000000000000001FE000001C18000000000000004000000000180000000000000180000000001800000000000002000000000008000000000000060000000000080000000000000400000000001800000000000006F800000000180000000000000508000000001000000000000007040000000010000000000000027000000000300000000000000200001FFE00600000000000000200007003C3C000000000000001F003C0007E00000000000000003FFE0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001000000800000000000000000010000C1000000000000000000091801C1882300000000000000393CB3C19CF300000000000000212494412299000000000000003124B6C134D10000000000000018186300086100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000^FS^FO100,75^GB300,200,10,B,0^XZ";
				int size = strlen(buffer);
				printf("Created job: %d\n", job_id);

				if (cupsStartDestDocument(CUPS_HTTP_DEFAULT, raw, info,
					job_id, "test label", CUPS_FORMAT_TEXT, num_options, options, 1) == HTTP_STATUS_CONTINUE)
				{
					if (cupsWriteRequestData(CUPS_HTTP_DEFAULT, buffer, size) == HTTP_STATUS_CONTINUE)
					{
						if (cupsFinishDestDocument(CUPS_HTTP_DEFAULT, raw
							,
							info) == IPP_STATUS_OK)
							puts("Document send succeeded.");
						else
							printf("Document send failed: %s\n", cupsLastErrorString());
					}
					else
						printf("Document send failed: %s\n", cupsLastErrorString());
				}

			}
			else
				printf("Unable to create job: %s\n", cupsLastErrorString());
		}
	}

	cupsFreeDests(count, dests);

	return (0);
}`

OS details
PRETTY_NAME="Raspbian GNU/Linux 9 (stretch)"
NAME="Raspbian GNU/Linux"
VERSION_ID="9"
VERSION="9 (stretch)"
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"

I am using CUPS 2.2.5

Regards
Dave

**Comment by michaelrsweet** [comment]
@ProgrammingPescador  Will investigate, but is there a reason why you don't have the queue setup with the driver for this printer? Even if you normally send raw (printer ready) files for printing, using the PPD provides the capability information that the CUPS API looks for...

**Comment by apostrophesoftware** [comment]
@michaelrsweet After some Googling I have been able to get this working using `job_id = cupsCreateJob(CUPS_HTTP_DEFAULT, raw->name, "Test Label", num_options, options)` to create the job.

The printer is setup as Zebra printer, however during the installation I selected "Raw" as the model.  the reason for this was during initial testing using a Nodejs library I was unable to send the ZPL commands directly to the Zebra queue and have the label print correctly!

The scenario I am working with is: A cloud solution creates a label, the ZPL for the label is sent to a Raspberry Pi, the Raspberry Pi sends the ZPL directly to the printer.

As stated this is working in Node js, but we are porting to C to get a smaller code footprint and better speeds on the Pi

**Comment by michaelrsweet** [comment]
The `cupsCreateJob` function calls into `cupsCreateDestJob` on recent versions of CUPS, so that probably won't help forever.

If you specify the "raw" option or "application/vnd.cups-raw" document format when you create the job, then the appropriate Zebra driver will work with preformatted label files.

Will continue investigating...

**Comment by michaelrsweet** [comment]
Confirmed that cupsCreateDestJob (more specifically cupsCopyDestInfo) wasn't working with raw queues.

**Comment by michaelrsweet** [comment]
[master 22974c5f3] The `cupsCopyDestInfo` function did not work with all print queues (Issue #5235)

**Comment by apostrophesoftware** [comment]
Hi

Thanks for the reply.

I have tried to use the following code to print the "raw" label
data: ipp_status_t status = cupsStartDocument(CUPS_HTTP_DEFAULT, raw->name,
job_id, "TestLabel", CUPS_FORMAT_RAW, lastdocument);
the return code is 100 decimal, and the error message returned bu
cupsLastErrorString is successful-ok.

The document does not print as if (status == IPP_STATUS_OK) fails and
therefore does not call cupsWriteRequestData to write the data.

I have looked in ipp.h and cannot find a constant that is equal to 100.
What is this error?

Regards

David Dwyer
Apostrophe Software Solutions Ltd
+44 (0)1728 648688
+44 (0)7544 250764
